### PR TITLE
LPS-178043 Make store file names globally unique

### PR DESCRIPTION
**Pull request by adolfopa**
# What is this about?

https://issues.liferay.com/browse/LPS-178043

As a first step towards implementing logical deletion, we need to make store file names globally unique. This will prevent updates to a file version from overwriting the previous contents in the store.

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

**Comment by adolfopa** [comment]
I've marked this as a draft as it depends on https://github.com/brianchandotcom/liferay-portal/pull/131637. Once that's merged, I'll rebase and tweak the validation methods a little bit.

**Comment by adolfopa** [comment]
ci:test:sf

**Comment by adolfopa** [comment]
ci:test:relevant

**Comment by adolfopa** [comment]
ci:reevaluate:1347401_3742

**Comment by adolfopa** [comment]
ci:test:sf

**Comment by adolfopa** [comment]
ci:test:relevant

**Comment by Jeremy-Ch** [comment]
ci:test:relevant

**Comment by Jeremy-Ch** [comment]
The failures are not related to this pr.

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/exportimport/data/handler/FileEntryStagedModelDataHandler.java [hunk 1]
```diff
@@ -939,6 +940,13 @@ private DDMFormValues _getImportDDMFormValues(
 		return DDMBeanTranslatorUtil.translate(ddmFormValues);
 	}
 
+	/**
+	 * @see com.liferay.portlet.documentlibrary.service.impl.DLFileEntryLocalServiceImplrviceImpl#_getStoreUUID(String)
+	 */
+	private String _getStoreUUID(String version) {
+		return version + StringPool.TILDE + UUID.randomUUID();
+	}
+
```
Comment (marcogalluzzi) [inline modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/exportimport/data/handler/FileEntryStagedModelDataHandler.java:949]: The class name linked with @see mention is not correct.

Also, I don't like having duplicated code: can we avoid this case?

**Review by marcogalluzzi — COMMENTED** [review]
I think you made some changes that may be confusing to anyone without a prior knowledge:

* StoreUUID: 
    * the name says this is an UUID, but now we are storing the concatenated string “version~UUID”
    * Storing version field also here seems redundant and may produce inconsistencies
* There are many methods handling files which require an input String parameter called “version” or “versionLabel”. Before we were passing the field “version”, which sounded correct, but now we are passing the field “StoreUUID” which seems unrelated.

I don't know how to solve it exactly, my suggestion:

* Create and use a new calculated field as you did before with `DLFileVersion.getStoreFileName` that concatenates fields version and storeUUID. I would only change the name to something like `getUniqueLabel`.

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/exportimport/data/handler/FileEntryStagedModelDataHandler.java (hunk 1 above)
Comment (marcogalluzzi) [reply modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/exportimport/data/handler/FileEntryStagedModelDataHandler.java:949]: > Change tracking requires the field to be stored in the database. That's why I removed the storeFileName computed value. There's no way out of that.

If the problem is the join with the CTSContent table, would it be something like this a viable solution?

* You could join tables by a concatenation, similarly to what is done in class `DLFileEntryTypeTableReferenceDefinition`: 
```java
			.innerJoinON(
				DLFileEntryTypeTable.INSTANCE,
				DLFileEntryTypeTable.INSTANCE.groupId.eq(
					DDMStructureTable.INSTANCE.groupId
				).and(
					DSLFunctionFactoryUtil.lower(
						DDMStructureTable.INSTANCE.structureKey
					).eq(
						DSLFunctionFactoryUtil.lower(
							DSLFunctionFactoryUtil.concat(
								new Scalar<>(
									DLUtil.getDDMStructureKey(
										StringPool.BLANK)),
								DLFileEntryTypeTable.INSTANCE.uuid))
					)
				)
			)
```

Comment (adolfopa) [reply modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/exportimport/data/handler/FileEntryStagedModelDataHandler.java:949]: I don't think so. This example embeds a literal value in the query; in our case, the join needs to be done with a value that needs to be computed for each row (the version number may be different for each row).

**Comment by adolfopa** [comment]
ci:test:relevant

**Review by marcogalluzzi — APPROVED** [review]

**Comment by marcogalluzzi** [comment]
ci:forward

**Comment by Jeremy-Ch** [comment]
@adolfopa Please help to resolve the conflicts, thanks.

**Comment by adolfopa** [comment]
Rebased and regenerated baseline commit; no other changes, so the previous review is still valid.

**Comment by adolfopa** [comment]
ci:forward

**Comment by adolfopa** [comment]
ci:forward

**Comment by Jeremy-Ch** [comment]
Hi @adolfopa Looks like this pr still has some conflicts when I tried to send this pr to BChan. Maybe it conflicts with https://github.com/brianchandotcom/liferay-portal/commit/b2d379a7b4afc4de4c5262a9cfd7e27fced510d2 I guess. Could you help to resolve it? I will send this pr manually after your updates. Thanks.

**Comment by adolfopa** [comment]
@Jeremy-Ch Conflicts resolved.

**Comment by Jeremy-Ch** [comment]
The failures are not related to this pr. Send to BChan manually https://github.com/brianchandotcom/liferay-portal/pull/132284

**Pull request closed without merging**